keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -638,24 +640,185 @@ always dispatched immediately.  Following is how the other requests
+| node-high       |  40 |  25 |   400 |
+| system          |  30 |  33 |   500 |
+| workload-high   |  40 |  50 |   833 |
+| workload-low    | 100 |  90 |  9000 |
+| global-default  |  20 |  50 |  9900 |
+| catch-all       |   5 |   0 | 10000 |
+
+Each non-exempt priority level `i` has two concurrency limits: its
+NominalConcurrencyLimit (`NominalCL(i)`) as defined above by
+configuration, and a CurrentConcurrencyLimit (`CurrentCL(i)`) that is
+used in dispatching requests.  The CurrentCLs are adjusted
+periodically, based on configuration, the current situation at
+adjustment time, and recent observations.  The "borrowing" resides in
+the differences between CurrentCL and NominalCL.  There is a lower
+bound on each non-exempt priority level's CurrentCL: `MinCL(i) =
+NominalCL(i) - BorrowableCL(i)`; the upper limit is imposed only by
+how many seats are available for borrowing from other priority levels.
+The sum of the CurrentCLs is always equal to the server's concurrency
+limit (ServerCL) plus or minus a little for rounding in the adjustment
+algorithm below.
 
 Dispatching is done independently for each priority level.  Whenever
-(1) a non-exempt priority level's number of running requests is zero
-or below the level's assured concurrency value and (2) that priority
-level has a non-empty queue, it is time to dispatch another request
+(1) a non-exempt priority level's number of occupied seats is zero or
+below the level's CurrentCL and (2) that priority level has a
+non-empty queue, it is time to consider dispatching another request
 for service.  The Fair Queuing for Server Requests algorithm below is
 used to pick a non-empty queue at that priority level.  Then the
-request at the head of that queue is dispatched.
+request at the head of that queue is dispatched if possible.
+
+Every 10 seconds, all the CurrentCLs are adjusted.  We do smoothing on
+the inputs to the adjustment logic in order to dampen control
+gyrations, in a way that lets a priority level reclaim lent seats at
+the nearest adjustment time.  The adjustments take into account the
+high watermark `HighSD(i)`, time-weighted average `AvgSD(i)`, and
+time-weighted population standard deviation `StDevSD(i)` of each
+priority level `i`'s seat demand over the just-concluded adjustment
+period.  A priority level's seat demand at any given moment is the sum
+of its occupied seats and the number of seats in the queued requests.
+We also define `EnvelopeSD(i) = AvgSD(i) + StDevSD(i)`.  The
+adjustment logic is driven by a quantity called smoothed seat demand
+(`SmoothSD(i)`), which does an exponential averaging of EnvelopeSD
+values using a coeficient A in the range (0,1) and immediately tracks
+EnvelopeSD when it exceeds SmoothSD.  The rule for updating priority
+level `i`'s SmoothSD at the end of an adjustment period is
+`SmoothSD(i) := max( EnvelopeSD(i), A*SmoothSD(i) + (1-A)*Envelope(i)
+)`.  The command line flag `--seat-demand-history-fraction` with a
```

Comment:
a cluster-admin doesn't seem likely to choose a better `A` than we will

Reply:
That's what default values are for?

Reply:
Also, I am not sure what value to use.  I made a rough guess, but it is only that.